File: Python/Project_Euler_S17.py
# Project Euler – Problem 17
# Title: Number Letter Counts
# Link: https://projecteuler.net/problem=17
#
# Count the number of letters needed
# to write numbers in words within a range.

ones = {
    0: "",
    1: "one", 2: "two", 3: "three", 4: "four", 5: "five",
    6: "six", 7: "seven", 8: "eight", 9: "nine",
    10: "ten", 11: "eleven", 12: "twelve", 13: "thirteen",
    14: "fourteen", 15: "fifteen", 16: "sixteen",
    17: "seventeen", 18: "eighteen", 19: "nineteen"
}

tens = {
    20: "twenty", 30: "thirty", 40: "forty",
    50: "fifty", 60: "sixty", 70: "seventy",
    80: "eighty", 90: "ninety"
}

def number_to_words(n):
    if n < 20:
        return ones[n]

    if n < 100:
        return tens[n // 10 * 10] + ones[n % 10]

    if n < 1000:
        if n % 100 == 0:
            return ones[n // 100] + "hundred"
        else:
            return ones[n // 100] + "hundredand" + number_to_words(n % 100)

    if n == 1000:
        return "onethousand"

total_letters = 0
for i in range(1, 1001):
    word = number_to_words(i)
    total_letters += len(word)

print("from 1 to 1000 inclusive were written out in words, number of letters would be used: ", total_letters)

File: C/Project_Euler_S17.c
// Project Euler – Problem 17
// Title: Number Letter Counts
// Link: https://projecteuler.net/problem=17
//
// Count the number of letters needed
// to write numbers in words within a range.

#include <stdio.h>
#include <string.h>

const char *ones[] = {
    "", "one", "two", "three", "four", "five",
    "six", "seven", "eight", "nine", "ten",
    "eleven", "twelve", "thirteen", "fourteen",
    "fifteen", "sixteen", "seventeen", "eighteen",
    "nineteen"
};

const char *tens[] = {
    "", "", "twenty", "thirty", "forty",
    "fifty", "sixty", "seventy", "eighty", "ninety"
};

void number_to_words(int n, char *result) {
    result[0] = '\0';

    if (n < 20) {
        strcat(result, ones[n]);
    }
    else if (n < 100) {
        strcat(result, tens[n / 10]);
        strcat(result, ones[n % 10]);
    }
    else if (n < 1000) {
        strcat(result, ones[n / 100]);
        strcat(result, "hundred");

        if (n % 100 != 0) {
            strcat(result, "and");
            char temp[50];
            number_to_words(n % 100, temp);
            strcat(result, temp);
        }
    }
    else if (n == 1000) {
        strcat(result, "onethousand");
    }
}

int main() {
    int total_letters = 0;
    char word[100];

    for (int i = 1; i <= 1000; i++) {
        number_to_words(i, word);
        total_letters += strlen(word);
    }

    printf("From 1 to 1000 inclusive were written out in words, number of letters would be used: %d\n",
           total_letters);

    return 0;
}
